proposals/0000-package-internals.rst

```diff
@@ -0,0 +1,148 @@
+   not intended for the public API of ``pack``. Yet the author of ``pack``
+   still wants to test functions exported by these internal modules.
+
+   The author currently has two options, both bad:
+
+   1. Expose the internal modules, perhaps naming them with ``Internal`` or
+      similar.
+
+   2. Structure the project to recompile the entire package in order to run
+      the testsuite, instead of using the already-compiled package.
+
+   This proposal allows the modules to remain internal while still being
+   available for testing without recompilation.
+
+2. Suppose package ``pack`` exports a very general interface to a particular
+   algorithm. Because it is so general, ``pack`` offers no way of, say, using
+   real-world data to seed the algorithm. Instead, downstream packages such as
+   ``pack-json`` and ``pack-csv`` provide such access.
+
+   If an internal module in ``pack`` needs testing, then there is no choice but
+   to expose the internal module, so it can be tested in a downstream package.
+
+3. Suppose two packages ``pack1`` and ``pack2`` are developed together, with
+   ``pack2`` depending on ``pack1``. Because the two packages form logically one
+   whole (but are separated because, say, ``pack1`` has many fewer dependencies
+   than ``pack2`` or other reasons), ``pack2`` needs to import internal modules
+   from ``pack1``. Yet these internal modules are not intended to form part of
+   any public API.
+
+   This is the case within GHC's boot libraries, where the ``ghc-prim`` package
+   is not intended to have a public API. Yet users cannot easily learn that
+   ``ghc-prim`` is essentially an implementation detail and should instead depend
+   only on ``base`` (which re-exports much, but not all, of ``ghc-prim``). This
+   proposal is intended to protect users from unknowingly depending on ``ghc-prim``
+   modules, which might change frequently.
+
+Proposed Change Specification
+-----------------------------
+
+1. Add a new command-line flag ``-expose-package-internals <pkg>``, where ``<pkg>`` is understood
+   in the same way as the argument to the existing ``-package`` flag.
+
+#. It is an error for the ``<pkg>`` named in ``-expose-package-internals`` not to be exposed.
+
+#. Imports of modules from a package named in ``-expose-package-internals`` skip the exposed-module check.
+
+#. The error GHC produces when a non-exposed module is imported warns the user about
+   exposing internal modules, with something like ::
+
+    Could not load module 'Data.OldList';
+    it is a hidden module in the package 'base-4.15.1.0'.
+    If you wish to be able to access an internal module anyway, you may
+    specify '-expose-package-internals base-4.15.1.0'; however, be warned
+    that the API from internal modules may change between minor version releases.
+
+#. The `Haskell Package Versioning Policy <https://pvp.haskell.org/>`_ specifies that changes in
+   internal modules do *not* require a major version change.
```

Comment:
This proposal makes API dependent on GHC compilation flags, which is a huge deviation from PVP. Don't we have enough fragiltiy in the ecosystem already?